File: internal/channel/router.go
package channel

import (
	"context"
	"fmt"
	"html"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/Z-M-Huang/openhive/internal/domain"
	"github.com/Z-M-Huang/openhive/internal/ws"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

const defaultMaxMessageLength = 10000

// Router implements domain.MessageRouter, connecting messaging channels to the
// main assistant via WebSocket. Inbound messages create Tasks and dispatch them
// via WSHub. Task results are routed back to the originating channel.
type Router struct {
	channels         map[string]domain.ChannelAdapter
	wsHub            domain.WSHub
	taskStore        domain.TaskStore
	sessionStore     domain.SessionStore
	messageStore     domain.MessageStore
	transactor       domain.Transactor
	logger           *slog.Logger
	mu               sync.RWMutex
	mainTeamID       string
	mainAssistantAID string
	maxMessageLength int
}

// RouterConfig holds configuration for the message router.
type RouterConfig struct {
	WSHub            domain.WSHub
	TaskStore        domain.TaskStore
	SessionStore     domain.SessionStore
	MessageStore     domain.MessageStore
	Transactor       domain.Transactor
	Logger           *slog.Logger
	MainTeamID       string
	MainAssistantAID string
	MaxMessageLength int
}

// NewRouter creates a new message router.
func NewRouter(cfg RouterConfig) *Router {
	maxLen := cfg.MaxMessageLength
	if maxLen <= 0 {
		maxLen = defaultMaxMessageLength
	}
	return &Router{
		channels:         make(map[string]domain.ChannelAdapter),
		wsHub:            cfg.WSHub,
		taskStore:        cfg.TaskStore,
		sessionStore:     cfg.SessionStore,
		messageStore:     cfg.MessageStore,
		transactor:       cfg.Transactor,
		logger:           cfg.Logger,
		mainTeamID:       cfg.MainTeamID,
		mainAssistantAID: cfg.MainAssistantAID,
		maxMessageLength: maxLen,
	}
}

// RegisterChannel registers a channel adapter, keyed by its JID prefix.
func (r *Router) RegisterChannel(adapter domain.ChannelAdapter) error {
	prefix := adapter.GetJIDPrefix()
	if prefix == "" {
		return &domain.ValidationError{Field: "prefix", Message: "channel prefix must not be empty"}
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.channels[prefix]; exists {
		return &domain.ConflictError{
			Resource: "channel",
			Message:  fmt.Sprintf("channel with prefix %q is already registered", prefix),
		}
	}

	// Wire up the onMessage callback
	adapter.OnMessage(func(jid string, content string) {
		if err := r.RouteInbound(jid, content); err != nil {
			r.logger.Error("failed to route inbound message", "jid", jid, "error", err)
		}
	})

	r.channels[prefix] = adapter
	r.logger.Info("channel registered", "prefix", prefix)
	return nil
}

// UnregisterChannel removes a channel adapter by its prefix.
func (r *Router) UnregisterChannel(prefix string) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.channels[prefix]; !exists {
		return &domain.NotFoundError{Resource: "channel", ID: prefix}
	}

	delete(r.channels, prefix)
	r.logger.Info("channel unregistered", "prefix", prefix)
	return nil
}

// RouteInbound handles an incoming message from a channel. Creates a Task,
// persists it, and dispatches to the main assistant container via WebSocket.
// It wraps task creation and session cursor update in a single transaction so
// that a crash mid-operation leaves no partial state.
func (r *Router) RouteInbound(jid string, content string) error {
	ctx := context.Background()

	// Enforce message length limit to prevent abuse.
	if len(content) > r.maxMessageLength {
		return &domain.ValidationError{
			Field:   "content",
			Message: fmt.Sprintf("message exceeds maximum length of %d characters", r.maxMessageLength),
		}
	}

	// Get or create session
	session, err := r.getOrCreateSession(ctx, jid)
	if err != nil {
		return fmt.Errorf("session error: %w", err)
	}

	// Escape content to prevent XML injection in the agent prompt.
	escapedContent := html.EscapeString(content)
	channelType := extractPrefix(jid)
	formatted := FormatUserMessage(channelType, escapedContent)

	// Create task
	taskID := uuid.NewString()
	now := time.Now()
	task := &domain.Task{
		ID:        taskID,
		TeamSlug:  "main",
		AgentAID:  session.AgentAID,
		JID:       jid,
		Status:    domain.TaskStatusPending,
		Prompt:    formatted,
		CreatedAt: now,
		UpdatedAt: now,
	}

	// Wrap task creation + session cursor update in a transaction.
	// If either fails, both are rolled back.
	var persistErr error
	if r.transactor != nil {
		taskStoreWithTx, ok := r.taskStore.(*TaskStoreWithTx)
		sessionStoreWithTx, ok2 := r.sessionStore.(*SessionStoreWithTx)
		if ok && ok2 {
			persistErr = r.transactor.WithTransaction(func(tx *gorm.DB) error {
				if err := taskStoreWithTx.CreateWithTx(tx, task); err != nil {
					return fmt.Errorf("failed to create task in transaction: %w", err)
				}
				session.LastTimestamp = now
				if err := sessionStoreWithTx.UpsertWithTx(tx, session); err != nil {
					return fmt.Errorf("failed to update session in transaction: %w", err)
				}
				return nil
			})
		} else {
			// Fall back to non-transactional path if concrete types not available.
			// Log a warning because the task and session update are no longer atomic —
			// a crash between the two calls could leave partial state.
			r.logger.Warn("transactor set but stores do not implement WithTx; falling back to non-transactional path",
				"task_store_type", fmt.Sprintf("%T", r.taskStore),
				"session_store_type", fmt.Sprintf("%T", r.sessionStore),
			)
			persistErr = r.taskStore.Create(ctx, task)
			if persistErr == nil {
				session.LastTimestamp = now
				if err := r.sessionStore.Upsert(ctx, session); err != nil {
					r.logger.Warn("failed to update session in fallback path", "jid", jid, "error", err)
				}
			}
		}
	} else {
		persistErr = r.taskStore.Create(ctx, task)
		if persistErr == nil {
			session.LastTimestamp = now
			if err := r.sessionStore.Upsert(ctx, session); err != nil {
				r.logger.Warn("failed to update session", "jid", jid, "error", err)
			}
		}
	}
	if persistErr != nil {
		return fmt.Errorf("failed to persist task: %w", persistErr)
	}

	// Persist inbound message
	if r.messageStore != nil {
		msg := &domain.Message{
			ID:        uuid.NewString(),
			ChatJID:   jid,
			Role:      "user",
			Content:   content,
			Timestamp: now,
		}
		if err := r.messageStore.Create(ctx, msg); err != nil {
			r.logger.Warn("failed to persist inbound message", "jid", jid, "error", err)
		}
	}

	// Dispatch via WebSocket
	dispatchMsg := ws.TaskDispatchMsg{
		TaskID:    taskID,
		AgentAID:  session.AgentAID,
		Prompt:    formatted,
		SessionID: session.SessionID,
	}

	encoded, err := ws.EncodeMessage(ws.MsgTypeTaskDispatch, dispatchMsg)
	if err != nil {
		return fmt.Errorf("failed to encode task dispatch: %w", err)
	}

	if err := r.wsHub.SendToTeam(r.mainTeamID, encoded); err != nil {
		r.logger.Warn("failed to send task dispatch (container may not be connected)", "team_id", r.mainTeamID, "error", err)
		// Don't return error - the task is persisted and can be retried
	}

	r.logger.Info("message routed inbound",
		"jid", jid,
		"task_id", taskID,
		"channel", channelType,
	)

	return nil
}

// RouteOutbound sends a response to the correct channel based on JID prefix.
func (r *Router) RouteOutbound(jid string, content string) error {
	r.logger.Debug("routing outbound", "jid", jid, "content_len", len(content))
	prefix := extractPrefix(jid)

	r.mu.RLock()
	adapter, exists := r.channels[prefix]
	r.mu.RUnlock()

	if !exists {
		return &domain.NotFoundError{Resource: "channel", ID: prefix}
	}

	// Strip any XML wrapper tags from agent response
	cleaned := StripResponseTags(content)

	if err := adapter.SendMessage(jid, cleaned); err != nil {
		return fmt.Errorf("failed to send message to channel %s: %w", prefix, err)
	}

	r.logger.Info("message routed outbound", "jid", jid, "channel", prefix)
	return nil
}

// GetChannels returns a map of registered channel prefixes to their connection status.
func (r *Router) GetChannels() map[string]bool {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make(map[string]bool, len(r.channels))
	for prefix, adapter := range r.channels {
		result[prefix] = adapter.IsConnected()
	}
	return result
}

// HandleTaskResult processes a task result received from a container.
// Updates the task in the DB and routes the response to the originating channel.
func (r *Router) HandleTaskResult(ctx context.Context, result *ws.TaskResultMsg) error {
	r.logger.Debug("handling task result",
		"task_id", result.TaskID,
		"status", result.Status,
		"has_result", result.Result != "",
		"has_error", result.Error != "",
	)

	// Update task in DB
	task, err := r.taskStore.Get(ctx, result.TaskID)
	if err != nil {
		return fmt.Errorf("task not found: %w", err)
	}

	now := time.Now()
	task.UpdatedAt = now
	task.CompletedAt = &now

	if result.Status == "completed" {
		task.Status = domain.TaskStatusCompleted
		task.Result = result.Result
	} else {
		task.Status = domain.TaskStatusFailed
		task.Error = result.Error
	}

	if err := r.taskStore.Update(ctx, task); err != nil {
		r.logger.Error("failed to update task", "task_id", result.TaskID, "error", err)
	}

	// Update session with agent's response timestamp
	jid := task.JID
	if jid != "" {
		session, sessionErr := r.sessionStore.Get(ctx, jid)
		if sessionErr == nil {
			session.LastAgentTimestamp = now
			if err := r.sessionStore.Upsert(ctx, session); err != nil {
				r.logger.Warn("failed to update session after result", "error", err)
			}
		}

		// Persist outbound message
		if r.messageStore != nil {
			var content string
			if result.Status == "completed" {
				content = result.Result
			}
			if content != "" {
				msg := &domain.Message{
					ID:        uuid.NewString(),
					ChatJID:   jid,
					Role:      "assistant",
					Content:   content,
					Timestamp: now,
				}
				if err := r.messageStore.Create(ctx, msg); err != nil {
					r.logger.Warn("failed to persist outbound message", "task_id", result.TaskID, "error", err)
				}
			}
		}

		// Route response to the originating channel
		var content string
		if result.Status == "completed" && result.Result != "" {
			content = result.Result
		} else if result.Status == "failed" {
			// Log the internal error for debugging but never expose it to the user.
			r.logger.Error("task failed", "task_id", result.TaskID, "internal_error", result.Error)
			content = "Sorry, I encountered an issue processing your request. Please try again."
		}
		if content != "" {
			if routeErr := r.RouteOutbound(jid, content); routeErr != nil {
				r.logger.Error("failed to route response", "task_id", result.TaskID, "error", routeErr)
				return routeErr
			}
		}
	}

	return nil
}

// RecoverInFlight re-dispatches tasks that were in-flight when the server crashed.
// It detects these by comparing lastTimestamp > lastAgentTimestamp on sessions.
func (r *Router) RecoverInFlight(ctx context.Context) error {
	sessions, err := r.sessionStore.ListAll(ctx)
	if err != nil {
		return fmt.Errorf("failed to list sessions for recovery: %w", err)
	}

	recovered := 0
	for _, session := range sessions {
		if session.LastTimestamp.After(session.LastAgentTimestamp) {
			// This session has an in-flight message. Find the pending/running task.
			tasks, err := r.taskStore.ListByTeam(ctx, "main")
			if err != nil {
				r.logger.Warn("recovery: failed to list tasks", "jid", session.ChatJID, "error", err)
				continue
			}
			for _, task := range tasks {
				if task.JID == session.ChatJID &&
					(task.Status == domain.TaskStatusPending || task.Status == domain.TaskStatusRunning) {
					// Re-dispatch the task
					dispatchMsg := ws.TaskDispatchMsg{
						TaskID:    task.ID,
						AgentAID:  session.AgentAID,
						Prompt:    task.Prompt,
						SessionID: session.SessionID,
					}
					encoded, encErr := ws.EncodeMessage(ws.MsgTypeTaskDispatch, dispatchMsg)
					if encErr != nil {
						r.logger.Warn("recovery: failed to encode dispatch", "task_id", task.ID, "error", encErr)
						continue
					}
					if sendErr := r.wsHub.SendToTeam(r.mainTeamID, encoded); sendErr != nil {
						r.logger.Warn("recovery: failed to re-dispatch task", "task_id", task.ID, "error", sendErr)
					} else {
						r.logger.Info("recovery: re-dispatched in-flight task", "task_id", task.ID, "jid", session.ChatJID)
						recovered++
					}
					break
				}
			}
		}
	}

	if recovered > 0 {
		r.logger.Info("in-flight recovery complete", "recovered", recovered)
	}
	return nil
}

func (r *Router) getOrCreateSession(ctx context.Context, jid string) (*domain.ChatSession, error) {
	session, err := r.sessionStore.Get(ctx, jid)
	if err != nil {
		// Create new session with main assistant as the default agent
		session = &domain.ChatSession{
			ChatJID:     jid,
			ChannelType: extractPrefix(jid),
			AgentAID:    r.mainAssistantAID,
		}
		if err := r.sessionStore.Upsert(ctx, session); err != nil {
			return nil, fmt.Errorf("failed to create session: %w", err)
		}
	}
	return session, nil
}

// FormatUserMessage wraps user content in a safe JSON-like format for agent consumption.
// Content is already HTML-escaped before being passed here.
func FormatUserMessage(channelType, escapedContent string) string {
	return fmt.Sprintf(`<user_message channel=%q>%s</user_message>`, channelType, escapedContent)
}

// StripResponseTags removes XML wrapper tags from agent responses.
func StripResponseTags(content string) string {
	// Remove <agent_response ...> wrapper if present
	if idx := strings.Index(content, ">"); idx != -1 && strings.HasPrefix(content, "<agent_response") {
		content = content[idx+1:]
		if endIdx := strings.LastIndex(content, "</agent_response>"); endIdx != -1 {
			content = content[:endIdx]
		}
	}
	return strings.TrimSpace(content)
}

// extractPrefix gets the channel prefix from a JID (everything before the first colon).
func extractPrefix(jid string) string {
	if idx := strings.Index(jid, ":"); idx != -1 {
		return jid[:idx]
	}
	return jid
}

// TaskStoreWithTx extends a TaskStore with transactional write support.
// This avoids adding WithTx to the domain.TaskStore interface.
type TaskStoreWithTx struct {
	domain.TaskStore
	impl interface{ CreateWithTx(tx *gorm.DB, task *domain.Task) error }
}

// CreateWithTx delegates to the concrete implementation.
func (s *TaskStoreWithTx) CreateWithTx(tx *gorm.DB, task *domain.Task) error {
	return s.impl.CreateWithTx(tx, task)
}

// SessionStoreWithTx extends a SessionStore with transactional write support.
type SessionStoreWithTx struct {
	domain.SessionStore
	impl interface{ UpsertWithTx(tx *gorm.DB, session *domain.ChatSession) error }
}

// UpsertWithTx delegates to the concrete implementation.
func (s *SessionStoreWithTx) UpsertWithTx(tx *gorm.DB, session *domain.ChatSession) error {
	return s.impl.UpsertWithTx(tx, session)
}

File: backend/src/channel/router.ts
/**
 * OpenHive Backend - Message Router
 *
 * Connects messaging channel adapters to the main assistant via WebSocket.
 * Inbound messages create Tasks and dispatch them to the main team container.
 * Task results are routed back to the originating channel.
 *
 * Design notes:
 *   - No transactions — task creation and session upsert are sequential
 *     (best-effort).
 *   - HTML escaping is implemented inline (no standard library equivalent).
 *   - Node.js is single-threaded so plain Map needs no locking.
 */

import { v4 as uuidv4 } from 'uuid';

import type { ChannelAdapter, MessageRouter, WSHub, TaskStore, SessionStore, MessageStore } from '../domain/interfaces.js';
import { ConflictError, NotFoundError, ValidationError } from '../domain/errors.js';
import type { Task, ChatSession, Message } from '../domain/types.js';
import { encodeMessage } from '../ws/protocol.js';
import { MsgTypeTaskDispatch } from '../ws/messages.js';
import type { TaskDispatchMsg, TaskResultMsg } from '../ws/messages.js';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_MAX_MESSAGE_LENGTH = 10000;

// ---------------------------------------------------------------------------
// Logger interface — minimal subset used internally
// ---------------------------------------------------------------------------

/**
 * Minimal logger interface required by Router.
 * Matches the shape of pino or any structured logger.
 */
export interface RouterLogger {
  debug(msg: string, ctx?: Record<string, unknown>): void;
  info(msg: string, ctx?: Record<string, unknown>): void;
  warn(msg: string, ctx?: Record<string, unknown>): void;
  error(msg: string, ctx?: Record<string, unknown>): void;
}

// ---------------------------------------------------------------------------
// RouterConfig
// ---------------------------------------------------------------------------

/**
 * Configuration for a Router instance.
 */
export interface RouterConfig {
  /** WebSocket hub used to dispatch task messages to the main container. */
  wsHub: WSHub;
  /** Persistence for tasks. */
  taskStore: TaskStore;
  /** Persistence for chat sessions. */
  sessionStore: SessionStore;
  /** Persistence for chat messages. May be null to disable message persistence. */
  messageStore: MessageStore | null;
  /** Structured logger. */
  logger: RouterLogger;
  /** TID of the main assistant container. */
  mainTeamID: string;
  /** Slug of the main team (used for task team_slug fields). Defaults to 'main'. */
  mainTeamSlug?: string;
  /** AID of the main assistant agent. */
  mainAssistantAID: string;
  /**
   * Maximum allowed inbound message length in characters.
   * Defaults to 10000 when not set or <= 0.
   */
  maxMessageLength?: number;
}

// ---------------------------------------------------------------------------
// HTML escape helper
// ---------------------------------------------------------------------------

/**
 * htmlEscape replaces the five characters that have special meaning in XML/HTML
 * with their safe entity equivalents: &, <, >, ", '.
 *
 * Used to prevent XML injection in the agent prompt wrapper.
 */
function htmlEscape(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&#34;')
    .replace(/'/g, '&#39;');
}

// ---------------------------------------------------------------------------
// Router
// ---------------------------------------------------------------------------

/**
 * Router connects messaging channel adapters to the main assistant via
 * WebSocket. It implements the MessageRouter interface.
 *
 * Inbound flow:
 *   Channel → RouteInbound → TaskStore.create → SessionStore.upsert
 *            → MessageStore.create (inbound msg) → WSHub.sendToTeam
 *
 * Outbound flow (from handleTaskResult):
 *   TaskStore.update → SessionStore.upsert → MessageStore.create (outbound)
 *   → RouteOutbound → ChannelAdapter.sendMessage
 *
 * Implements the MessageRouter interface.
 */
export class Router implements MessageRouter {
  private readonly channels: Map<string, ChannelAdapter> = new Map();
  private readonly wsHub: WSHub;
  private readonly taskStore: TaskStore;
  private readonly sessionStore: SessionStore;
  private readonly messageStore: MessageStore | null;
  private readonly logger: RouterLogger;
  private readonly mainTeamID: string;
  private readonly mainTeamSlug: string;
  private readonly mainAssistantAID: string;
  private readonly maxMessageLength: number;

  constructor(cfg: RouterConfig) {
    this.wsHub = cfg.wsHub;
    this.taskStore = cfg.taskStore;
    this.sessionStore = cfg.sessionStore;
    this.messageStore = cfg.messageStore;
    this.logger = cfg.logger;
    this.mainTeamID = cfg.mainTeamID;
    this.mainTeamSlug = cfg.mainTeamSlug ?? 'main';
    this.mainAssistantAID = cfg.mainAssistantAID;
    this.maxMessageLength =
      cfg.maxMessageLength !== undefined && cfg.maxMessageLength > 0
        ? cfg.maxMessageLength
        : DEFAULT_MAX_MESSAGE_LENGTH;
  }

  // ---------------------------------------------------------------------------
  // registerChannel
  // ---------------------------------------------------------------------------

  /**
   * Registers a channel adapter. The adapter is keyed by its JID prefix.
   * Wires up the onMessage callback so inbound messages are automatically routed.
   * Throws ConflictError if a channel with the same prefix is already registered.
   * Throws ValidationError if the prefix is empty.
   *
   * Implements MessageRouter.registerChannel.
   */
  async registerChannel(adapter: ChannelAdapter): Promise<void> {
    const prefix = adapter.getJIDPrefix();
    if (prefix === '') {
      throw new ValidationError('prefix', 'channel prefix must not be empty');
    }

    if (this.channels.has(prefix)) {
      throw new ConflictError(
        'channel',
        `channel with prefix "${prefix}" is already registered`,
      );
    }

    // Wire up the onMessage callback before registering.
    adapter.onMessage((jid: string, content: string) => {
      this.routeInbound(jid, content).catch((err: unknown) => {
        this.logger.error('failed to route inbound message', {
          jid,
          error: err instanceof Error ? err.message : String(err),
        });
      });
    });

    this.channels.set(prefix, adapter);
    this.logger.info('channel registered', { prefix });
  }

  // ---------------------------------------------------------------------------
  // unregisterChannel
  // ---------------------------------------------------------------------------

  /**
   * Removes a registered channel adapter by prefix.
   * Throws NotFoundError if no channel with that prefix is registered.
   *
   * Implements MessageRouter.unregisterChannel.
   */
  async unregisterChannel(prefix: string): Promise<void> {
    if (!this.channels.has(prefix)) {
      throw new NotFoundError('channel', prefix);
    }

    this.channels.delete(prefix);
    this.logger.info('channel unregistered', { prefix });
  }

  // ---------------------------------------------------------------------------
  // routeInbound
  // ---------------------------------------------------------------------------

  /**
   * Handles an inbound message from a channel. Steps:
   *   1. Enforce maximum message length.
   *   2. Get or create a ChatSession for the JID.
   *   3. HTML-escape the content and wrap in XML.
   *   4. Create a Task record.
   *   5. Persist task + update session timestamp (sequential, best-effort).
   *   6. Persist inbound Message (best-effort — failure is logged, not thrown).
   *   7. Dispatch via WebSocket (best-effort — failure is logged, not thrown).
   *
   * Implements MessageRouter.routeInbound.
   */
  async routeInbound(jid: string, content: string): Promise<void> {
    // (1) Enforce message length limit.
    if (content.length > this.maxMessageLength) {
      throw new ValidationError(
        'content',
        `message exceeds maximum length of ${this.maxMessageLength} characters`,
      );
    }

    // (2) Get or create session.
    const session = await this.getOrCreateSession(jid);

    // (3) Escape and format.
    const escapedContent = htmlEscape(content);
    const channelType = extractPrefix(jid);
    const formatted = formatUserMessage(channelType, escapedContent);

    // (4) Build task.
    const taskID = uuidv4();
    const now = new Date();
    const task: Task = {
      id: taskID,
      team_slug: this.mainTeamSlug,
      agent_aid: session.agent_aid,
      jid,
      status: 'pending',
      prompt: formatted,
      created_at: now,
      updated_at: now,
      completed_at: null,
    };

    // (5) Persist task, then update session timestamp sequentially.
    await this.taskStore.create(task);

    session.last_timestamp = now;
    try {
      await this.sessionStore.upsert(session);
    } catch (err) {
      this.logger.warn('failed to update session after task creation', {
        jid,
        error: err instanceof Error ? err.message : String(err),
      });
    }

    // (6) Persist inbound message (best-effort).
    if (this.messageStore !== null) {
      const msg: Message = {
        id: uuidv4(),
        chat_jid: jid,
        role: 'user',
        content,
        timestamp: now,
      };
      try {
        await this.messageStore.create(msg);
      } catch (err) {
        this.logger.warn('failed to persist inbound message', {
          jid,
          error: err instanceof Error ? err.message : String(err),
        });
      }
    }

    // (7) Dispatch via WebSocket (best-effort).
    const dispatchMsg: TaskDispatchMsg = {
      task_id: taskID,
      agent_aid: session.agent_aid ?? this.mainAssistantAID,
      prompt: formatted,
      session_id: session.session_id,
    };

    try {
      const encoded = encodeMessage(MsgTypeTaskDispatch, dispatchMsg);
      await this.wsHub.sendToTeam(this.mainTeamID, encoded);
    } catch (err) {
      this.logger.warn(
        'failed to send task dispatch (container may not be connected)',
        {
          team_id: this.mainTeamID,
          error: err instanceof Error ? err.message : String(err),
        },
      );
      // Not returned — task is persisted and can be recovered.
    }

    this.logger.info('message routed inbound', {
      jid,
      task_id: taskID,
      channel: channelType,
    });
  }

  // ---------------------------------------------------------------------------
  // routeOutbound
  // ---------------------------------------------------------------------------

  /**
   * Routes a response to the correct channel based on the JID prefix.
   * Strips <agent_response> XML wrapper tags before sending.
   * Throws NotFoundError if no channel is registered for the JID prefix.
   *
   * Implements MessageRouter.routeOutbound.
   */
  async routeOutbound(jid: string, content: string): Promise<void> {
    this.logger.debug('routing outbound', { jid, content_len: content.length });

    const prefix = extractPrefix(jid);
    const adapter = this.channels.get(prefix);
    if (adapter === undefined) {
      throw new NotFoundError('channel', prefix);
    }

    const cleaned = stripResponseTags(content);
    await adapter.sendMessage(jid, cleaned);

    this.logger.info('message routed outbound', { jid, channel: prefix });
  }

  // ---------------------------------------------------------------------------
  // getChannels
  // ---------------------------------------------------------------------------

  /**
   * Returns a map of registered channel prefixes to their connection status.
   * Implements MessageRouter.getChannels.
   */
  getChannels(): Record<string, boolean> {
    const result: Record<string, boolean> = {};
    for (const [prefix, adapter] of this.channels.entries()) {
      result[prefix] = adapter.isConnected();
    }
    return result;
  }

  // ---------------------------------------------------------------------------
  // handleTaskResult
  // ---------------------------------------------------------------------------

  /**
   * Processes a task result received from a container via WebSocket.
   * Steps:
   *   1. Load task from DB and update its status/result/error.
   *   2. Update the session's lastAgentTimestamp.
   *   3. Persist outbound message (if completed with result content).
   *   4. Route response to originating channel.
   *      On task failure, send a generic user-facing message — NEVER expose
   *      internal error details.
   *
   * Implements MessageRouter.handleTaskResult.
   */
  async handleTaskResult(result: TaskResultMsg): Promise<void> {
    this.logger.debug('handling task result', {
      task_id: result.task_id,
      status: result.status,
      has_result: (result.result ?? '') !== '',
      has_error: (result.error ?? '') !== '',
    });

    // (1) Update task in DB.
    const task = await this.taskStore.get(result.task_id);
    const now = new Date();
    task.updated_at = now;
    task.completed_at = now;

    if (result.status === 'completed') {
      task.status = 'completed';
      task.result = result.result;
    } else {
      task.status = 'failed';
      task.error = result.error;
    }

    try {
      await this.taskStore.update(task);
    } catch (err) {
      this.logger.error('failed to update task', {
        task_id: result.task_id,
        error: err instanceof Error ? err.message : String(err),
      });
    }

    // (2) Update session with agent response timestamp.
    const jid = task.jid;
    if (jid === undefined || jid === '') {
      return;
    }

    try {
      const session = await this.sessionStore.get(jid);
      session.last_agent_timestamp = now;
      await this.sessionStore.upsert(session);
    } catch (err) {
      this.logger.warn('failed to update session after result', {
        error: err instanceof Error ? err.message : String(err),
      });
    }

    // (3) Persist outbound message (best-effort).
    if (this.messageStore !== null) {
      const outboundContent = result.status === 'completed' ? (result.result ?? '') : '';
      if (outboundContent !== '') {
        const msg: Message = {
          id: uuidv4(),
          chat_jid: jid,
          role: 'assistant',
          content: outboundContent,
          timestamp: now,
        };
        try {
          await this.messageStore.create(msg);
        } catch (err) {
          this.logger.warn('failed to persist outbound message', {
            task_id: result.task_id,
            error: err instanceof Error ? err.message : String(err),
          });
        }
      }
    }

    // (4) Route response to the originating channel.
    let responseContent: string;
    if (result.status === 'completed' && (result.result ?? '') !== '') {
      responseContent = result.result ?? '';
    } else if (result.status === 'failed') {
      // Log internal error but NEVER expose it to the user.
      this.logger.error('task failed', {
        task_id: result.task_id,
        internal_error: result.error ?? '',
      });
      responseContent = 'Sorry, I encountered an issue processing your request. Please try again.';
    } else {
      return;
    }

    try {
      await this.routeOutbound(jid, responseContent);
    } catch (err) {
      this.logger.error('failed to route response', {
        task_id: result.task_id,
        error: err instanceof Error ? err.message : String(err),
      });
      throw err;
    }
  }

  // ---------------------------------------------------------------------------
  // recoverInFlight
  // ---------------------------------------------------------------------------

  /**
   * Re-dispatches tasks that were in-flight when the server crashed.
   * Detects in-flight sessions by comparing last_timestamp > last_agent_timestamp.
   * For each such session, finds the pending/running task and re-dispatches it.
   *
   * Implements MessageRouter.recoverInFlight.
   */
  async recoverInFlight(): Promise<void> {
    let sessions: ChatSession[];
    try {
      sessions = await this.sessionStore.listAll();
    } catch (err) {
      throw new Error(
        `failed to list sessions for recovery: ${err instanceof Error ? err.message : String(err)}`,
      );
    }

    let recovered = 0;

    for (const session of sessions) {
      if (session.last_timestamp <= session.last_agent_timestamp) {
        continue;
      }

      // This session has an in-flight message — find the pending/running task.
      let tasks: Task[];
      try {
        tasks = await this.taskStore.listByTeam(this.mainTeamSlug);
      } catch (err) {
        this.logger.warn('recovery: failed to list tasks', {
          jid: session.chat_jid,
          error: err instanceof Error ? err.message : String(err),
        });
        continue;
      }

      for (const task of tasks) {
        if (
          task.jid === session.chat_jid &&
          (task.status === 'pending' || task.status === 'running')
        ) {
          // Re-dispatch this task.
          const dispatchMsg: TaskDispatchMsg = {
            task_id: task.id,
            agent_aid: session.agent_aid ?? this.mainAssistantAID,
            prompt: task.prompt,
            session_id: session.session_id,
          };

          try {
            const encoded = encodeMessage(MsgTypeTaskDispatch, dispatchMsg);
            await this.wsHub.sendToTeam(this.mainTeamID, encoded);
            this.logger.info('recovery: re-dispatched in-flight task', {
              task_id: task.id,
              jid: session.chat_jid,
            });
            recovered++;
          } catch (err) {
            this.logger.warn('recovery: failed to re-dispatch task', {
              task_id: task.id,
              error: err instanceof Error ? err.message : String(err),
            });
          }

          // Only re-dispatch the first matching task per session.
          break;
        }
      }
    }

    if (recovered > 0) {
      this.logger.info('in-flight recovery complete', { recovered });
    }
  }

  // ---------------------------------------------------------------------------
  // Private helpers
  // ---------------------------------------------------------------------------

  /**
   * Gets an existing chat session for the JID, or creates a new one
   * with the main assistant as the default agent.
   * Returns the existing session or creates a new one with main assistant defaults.
   */
  private async getOrCreateSession(jid: string): Promise<ChatSession> {
    try {
      return await this.sessionStore.get(jid);
    } catch {
      // Session does not exist — create a new one.
      const session: ChatSession = {
        chat_jid: jid,
        channel_type: extractPrefix(jid),
        last_timestamp: new Date(0),
        last_agent_timestamp: new Date(0),
        agent_aid: this.mainAssistantAID,
      };
      await this.sessionStore.upsert(session);
      return session;
    }
  }
}

// ---------------------------------------------------------------------------
// Pure functions — exported for direct testing
// ---------------------------------------------------------------------------

/**
 * formatUserMessage wraps user content in an XML tag for agent consumption.
 * Content is already HTML-escaped before being passed here.
 *
 * @param channelType - The channel type prefix (e.g. "discord", "whatsapp").
 * @param escapedContent - HTML-escaped message content.
 * @returns XML-wrapped string.
 */
export function formatUserMessage(channelType: string, escapedContent: string): string {
  return `<user_message channel="${channelType}">${escapedContent}</user_message>`;
}

/**
 * stripResponseTags removes <agent_response> XML wrapper tags from agent
 * responses, returning only the inner content (trimmed).
 *
 * Handles both bare <agent_response> and tags with attributes
 * (e.g. <agent_response version="1">).
 *
 * @param content - Raw content from the agent, possibly wrapped in tags.
 * @returns Cleaned content with outer tags removed.
 */
export function stripResponseTags(content: string): string {
  // Remove opening <agent_response ...> tag if present.
  if (content.startsWith('<agent_response')) {
    const closeIdx = content.indexOf('>');
    if (closeIdx !== -1) {
      content = content.slice(closeIdx + 1);
      // Remove closing </agent_response> tag.
      const endIdx = content.lastIndexOf('</agent_response>');
      if (endIdx !== -1) {
        content = content.slice(0, endIdx);
      }
    }
  }
  return content.trim();
}

/**
 * extractPrefix gets the channel prefix from a JID (everything before the
 * first colon).
 *
 * Examples:
 *   "discord:123456"  → "discord"
 *   "whatsapp:+1234"  → "whatsapp"
 *   "nocolon"         → "nocolon"
 *
 * Exported for testing.
 */
export function extractPrefix(jid: string): string {
  const idx = jid.indexOf(':');
  if (idx !== -1) {
    return jid.slice(0, idx);
  }
  return jid;
}
